src/jsTyping/jsTyping.ts

```diff
@@ -264,6 +263,29 @@ namespace ts.JsTyping {
                 return;
             }
 
+            // Collect all toplevel package names if we're looking in node_modules

```

Comment:
Wait, isn't this redundant with `getTypingNamesFromJson` called at line 155? This makes me wonder why we had this function to crawl through `node_modules` in the first place.

Reply:
This may be a good moment to document the current behavior of `discoverTypings` a little more in comments, since I'm actually not sure what are all the places we look to discover dependency/typing information.

Reply:
Oh. I didn't even notice that.

Yeah, that would make this method redundant.

I can only guess at what was going through people's heads, but one bit where this particular method _may_ have made sense was when you did an `npm install` _without_ a package.json at all, which I feel like it shouldn't even be a scenario we bother with?

Reply:
This code is so old. (#6448) I'm also guessing at the original intentions, but I think it's safe to say we don't have to put too much stock in assumptions from 2016. 

Agree with not bothering with `node_modules` without a `package.json` (or `bower_components` without `bower.json` for that matter?)

One thing that sticks out to me is that this code tries to exclude dependent packages that bundle their own typings. Which is good. But I don't see how that would work since the same package would have been included upstream in `getTypingNamesFromJson` anyway. This may be a good part of the code to keep (and make it work)

Reply:
ahhhhh I get it now. `getTypingNamesFromJson` seems to be an initial pass at a "best guess", whereas we need to do this work _again_ if `node_modules` is populated in case those packages define their own `typings`/`types` field. We still need to do filtering here but I'll see if I can refactor it to be nicer. We definitely need to do what _both_ of these functions are doing:

1. `getTypingNamesFromJson` gives us a set of typing names to look up even if the user hasn't done `npm install` yes.
2. `getTypingNamesFromPackagesFolder` collects typing names for packages not in package.json/bower.json (unnecessary, imo), but _also_ checks for the `typings`/`types` field.

Maybe it's worth doing a small, well-documented rewrite here to put together these use cases into a single function that does both in a single pass?

Reply:
That would be so neat :)

I think we agree on the desired behavior (barring any surprises, since I haven't really read the existing code thoroughly).